### DISCUSSION TITLE
eWidgets with sprites?

### DISCUSSION THREAD

**Discussion by peteDDD** [Bodmer/TFT_eSPI#2116] [Q&A - Sprites] [answered]
Trying to use the GraphWidget and TraceWidget with a Sprite.
Here is my set-up below.  I am using img as my sprite with a pointer to tft.
Then I have GraphWidget (gr) with a pointer to img and TraceWidget (tr) with a pointer to the GraphWidget (gr)

```
TFT_eSPI tft = TFT_eSPI();
TFT_eSprite img = TFT_eSprite(&tft);
FT62XXTouchScreen touchScreen = FT62XXTouchScreen(TFT_WIDTH, PIN_SDA, PIN_SCL);
#include <TFT_eWidget.h> // Widget library
GraphWidget gr = GraphWidget(&img); // Graph widget gr instance with pointer 
TraceWidget tr = TraceWidget(&gr);  // Graph trace tr with pointer to gr
```

In setup(), I init the tft

```
  tft.init();
  tft.setRotation(1);

```

Nothing too crazy...
I do things like the following to create the Sprite and to build a graph (which all works fine without the sprite)

```
  img.setColorDepth(16);
  img.createSprite(480, 320);  // full screen
  img.fillSprite(TFT_TRANSPARENT);

  gr.createGraph(460, 300, tft.color565(5, 5, 5));
  gr.setGraphScale(gxLow, gxHigh, gyLow, gyHigh);  // YES, all of these variables are defined but I save you all that code
  gr.setGraphGrid(gxLow, 60.0, gyLow, 10.0, TFT_BLUE);
  gr.drawGraph(10, 10);
```

Then I push the sprite, but nothing is displayed... 

`img.pushSprite(10, 10, TFT_TRANSPARENT);`

Clearly (or should I say "Transparently") I am missing something fundamental here.  Anyone able to help me through this knothole?
Thanks!

**Comment by Bodmer** [comment, marked as answer]
Perhaps the sprite is not being created as there is insufficent RAM (~300kbytes needed). img.createSprite(480, 320); will return a nullptr if the creation fails.

**Reply by peteDDD** [reply]
Thanks for that very, very prompt reply.  That makes sense.  I will look at that.  

Is there a way to use PSRAM (which I have tons of) to hold the Sprite?

**Reply by Bodmer** [reply]
If you are using an ESP32 and you enable PSRAM in the IDE then that should work and the sprite will automatically be put in PSRAM.

To check for creation you can use code like this:
```
  if (img.createSprite(100, 100) == nullptr) Serial.println("Sprite NOT created, insufficient RAM!");
  else Serial.println("YES! Sprite created.");
```

**Reply by peteDDD** [reply]
BAM!  That was it.  I just had to turn on the PSRAM.  Half a day googling figuring it was my syntax...

Thanks for the help and thanks for a great library.
